matplotcheck/raster.py

```diff
@@ -100,68 +97,62 @@ def assert_legend_accuracy_classified_image(
-        assert legends, "No legend displayed"
+        assert list(im_data), "No Image Displayed"
 
         # Retrieve legend entries and find which element of all_label_options
         # matches that entry
-        legend_dict = {}
-        for p in [
-            p
-            for sublist in [leg.get_patches() for leg in legends]
-            for p in sublist
-        ]:
-            label = p.get_label().lower()
-            legend_dict[p.get_facecolor()] = self._which_label(
-                label, all_label_options
-            )
 
-        # Check that each legend entry label is in one of all_label_options
-        assert len([val for val in legend_dict.values() if val]) == len(
-            all_label_options
-        ), "Incorrect legend labels"
-
-        # Create two copies of image array, one filled with the plot data class
-        # labels (im_data_labels) and the other with the expected labels
-        # (im_expected_labels)
-        im_class_dict = {}
-        for val in np.unique(im_data):
-            im_class_dict[val] = legend_dict[im_cmap(im.norm(val))]
-        im_data_labels = [
-            [im_class_dict[val] for val in row] for row in im_data.data
-        ]
-        im_expected_labels = [
-            [all_label_options[val][0] for val in row] for row in im_expected
+        labels = self.get_legend_labels()
+
+        assert len(labels) == len(all_label_options), (
+            "Number of label options provided doesn't match the number of"
+            " labels found in the image."
+        )
+
+        labels_check = [
+            self._which_label(label, all_label_options[i])
+            for i, label in enumerate(labels)
         ]
 
-        # Check that expected and actual labels match up
+        # Check that each legend entry label is in one of all_label_options
+        assert all(
+            labels_check
+        ), "Provided legend labels don't match labels found."
+
+        # Check that expected and actual arrays data match up
         assert np.array_equal(
-            im_data_labels, im_expected_labels
-        ), "Incorrect legend to data relation"
+            im_data, im_expected
+        ), "Expected image data doesn't match data in image."
 
         # IMAGE TESTS/HELPER FUNCTIONS
 
+    def get_legend_labels(self):
+        """Return labels from legend in a list
+
+        Returns
+        -------
+        labels: List
+            List of labels found in the legend of a raster plot.
+        """
+
+        # Retrieve legend
+        legends = self.get_legends()
+        assert legends, "No legend displayed"
+
+        # Get each patch stored in the legends object
+        patches = [leg.get_patches() for leg in legends]
+
+        # Go through each patch to retrieve the labels from the legend
+        return [
+            label.get_label().lower()
+            for sublist in patches
+            for label in sublist
+        ]
+
     def get_plot_image(self):
```

Comment:
move this up - and then make sure all methods are defined before they are called in this module.